Add Azure KeyVault certificate storage and retrieval

twsouthwick opened the pull request:
Fixes #34

twsouthwick commented:
This is blocked right now by https://github.com/Azure/azure-sdk-for-net/issues/9963. Wanted to get any feedback now, so once a fix is available there, it can be used

natemcmaster commented on src/LetsEncrypt.Azure/AzureKeyVaultExtensions.cs:
```diff
@@ -0,0 +1,41 @@
+        /// <param name="builder">A LetsEncrypt service builder.</param>
+        /// <param name="config">Configuration for KeyVault connections.</param>
+        /// <returns>The original LetsEncrypt service builder.</returns>
+        public static ILetsEncryptServiceBuilder AddKeyVaultCertificateRepository(this ILetsEncryptServiceBuilder builder, Action<AzureKeyVaultCertificateRepositoryOptions>? config = null)
```
```suggestion
        public static ILetsEncryptServiceBuilder AddKeyVaultCertificateRepository(this ILetsEncryptServiceBuilder builder)
                    => builder.AddKeyVaultCertificateRepository(_ => {});
                   
        public static ILetsEncryptServiceBuilder AddKeyVaultCertificateRepository(this ILetsEncryptServiceBuilder builder, Action<AzureKeyVaultCertificateRepositoryOptions> config)
```

Nit: can we split this into two overloads? I've found that using optional parameters in the long run actually makes it more difficult to extend and API with new parameters in the future.

twsouthwick replied:
Sure, no problem

natemcmaster commented on src/LetsEncrypt/ICertificateSource.cs:
```diff
@@ -6,10 +6,18 @@
+    /// <summary>
+    /// Defines a source for certificates.
+    /// </summary>
+    public interface ICertificateSource
```
FYI - this is okay to make this public for now, but I will probably make some changes before this gets released to nuget.org. I want to finish cleaning up the implementation that uses the X509Store so that it unifies into the use of one interface for providing certificates that the server can use.

twsouthwick replied:
That's fine - I just needed to implement it.

natemcmaster commented on src/LetsEncrypt.Azure/AzureKeyVaultCertificateRepository.cs:
```diff
@@ -0,0 +1,143 @@
+            }
+        }
+
+        private async ValueTask<bool> ShouldImportVersionAsync(string domainName, X509Certificate2 certificate, CancellationToken token)
```
Maybe I missed something, but this is unlikely, right? Do we need this check? When would we be pulling a cert from Let's Encrypt that is already provided by Azure Key Vault?

twsouthwick replied:
The saving method is called even when it's pulled from a source. So, if it is pulled from keyvault, this check is to prevent it from being imported again when it's already there. So, maybe a better fix would be to only call the save method when new certificate is created from LetsEncrypt, but I put this in there as a stop gap.

natemcmaster commented on src/LetsEncrypt.Azure/AzureKeyVaultCertificateRepositoryOptions.cs:
```diff
@@ -0,0 +1,32 @@
+        [Url]
+        [Required]
+#endif
+        public string Url { get; set; } = null!;
```
```suggestion
        public string? Url { get; set; }
```
Does this work? Or what about something longer, but with an API that explains the URI's purpose?
```suggestion
        public Uri? AzureKeyVaultEndpoint { get; set; }
```

twsouthwick replied:
Updated

natemcmaster commented on src/LetsEncrypt.Azure/AzureKeyVaultExtensions.cs:
```diff
@@ -0,0 +1,41 @@
+        /// <param name="builder">A LetsEncrypt service builder.</param>
+        /// <param name="config">Configuration for KeyVault connections.</param>
+        /// <returns>The original LetsEncrypt service builder.</returns>
+        public static ILetsEncryptServiceBuilder AddKeyVaultCertificateRepository(this ILetsEncryptServiceBuilder builder, Action<AzureKeyVaultCertificateRepositoryOptions>? config = null)
```
Can we name this `. PersistCertificatesToKeyVault()`? I'm going for a naming pattern here that is similar to the data protection APIs and already have `.PersistCertificatesToDirectory()`

twsouthwick replied:
Since this both persists and retrieves from KeyVault, should it be separated to `.PersistCertifcatesToAzureKeyVault()` and `.AddAzureKeyVaultCertifcateStore()`?

twsouthwick commented on src/LetsEncrypt.Azure/McMaster.AspNetCore.LetsEncrypt.Azure.csproj:
```diff
@@ -0,0 +1,30 @@
+  </PropertyGroup>
+
+  <PropertyGroup Condition=" '$(TargetFramework)' == 'netcoreapp3.0' ">
+    <DefineConstants>$(DefineConstants);FEATURE_VALIDATE_DATA_ANNOTATIONS</DefineConstants>
```
The data annotations stuff wasn't added until ASP.NET Core 2.2. It uses the `Microsoft.Extensions.Options.DataAnnotations` package. It supports .NET Standard 2.0, but brings in v3.1.x of the options packages. If that's fine, I can enable it for .NET Standard

natemcmaster replied:
Thanks for clarifying. I like the approach you have. Let's leave it as-is.

twsouthwick commented:
I updated things according to most of the comments - I left some unresolved as I'm unsure if I resolved it completely.

- Regarding Azure/azure-sdk-for-net#9963, without it, importing fails. The fix is in, and will probably be available soon. You're call on merging this in and updating later when the fix is available
- I have a few unit tests in to ensure things are registered/resolved as expected. I also added a sample with steps for integration testing.

natemcmaster commented on src/LetsEncrypt.Azure/McMaster.AspNetCore.LetsEncrypt.Azure.csproj:
```diff
@@ -0,0 +1,30 @@
+
+  <ItemGroup>
+    <PackageReference Include="Azure.Identity" Version="1.1.1" />
+    <PackageReference Include="Azure.Security.KeyVault.Certificates" Version="4.0.0" />
```
Looks like the required fix for importing certs was released a few days ago. Now that this is out, let's move forward with this feature :D. I'm eager to get this merged and released.
```suggestion
    <PackageReference Include="Azure.Security.KeyVault.Certificates" Version="4.0.1" />
```

twsouthwick replied:
Cool!

natemcmaster commented:
> I updated things according to most of the comments 

By the way, I realize now I had a draft comment that I failed to send. Thank you for making the updates! This PR looks like it's in good shape and I think it's ready to go once we have a fix for Azure/azure-sdk-for-net#9963.

twsouthwick commented:
You mentioned you had a draft comment, but I don't see it anywhere. I've got the fixed version of the KeyVault library in and merges from master. Let me know if there was anything else.

natemcmaster commented:
I just meant that I wrote a comment 3 weeks ago after you made updates but missed the "submit review" step that actually sends it to you. The comment was basically "your changes look good, but let's wait for the azure sdk update."

Thanks for the contribution! This will go out to nuget.org in a few weeks.
